pkg/util/gvk-gvr-mapper.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GvkGvrMapper is a thread-safe mapping between GVKs and GVRs.
+type GvkGvrMapper interface {
+	// Add adds a mapping between the given GVK and GVR.
+	Add(gvk schema.GroupVersionKind, gvr schema.GroupVersionResource)
+	// DeleteByGvkKey deletes the mapping associated with the given GVK.
+	// The key format is that outputted by utils.KeyForGroupVersionKind.
+	DeleteByGvkKey(string)
```

Comment:
Why not make the entire interface use the upstream structure (instead of the local string encoding)?

Reply:
Add/delete of mappings is done by the crd handling code. For convenience, deletion is done by the key as string since that is what's present when used.

I agree it would be better if the delete functions used the upstream structure, but that would require either:
1. Writing a reverse function that extracts GVK from the key, while handling empty group and assuming validity.
2. Refactoring the crd handling code, and the mappings in the controller, which need to be done regardless because they suffer from possible race conditions.

I prefer option 2, which is a part of refactoring the placement code as a whole that will follow later.

What do you think?

Reply:
Sure, that would be OK.